src/poetry/repositories/pool.py

```diff
@@ -1,58 +1,66 @@
 from __future__ import annotations
 
+from collections import OrderedDict
+from enum import Enum
+from enum import auto
 from typing import TYPE_CHECKING
 
 from poetry.repositories.exceptions import PackageNotFound
-from poetry.repositories.repository import Repository
 
 
 if TYPE_CHECKING:
     from poetry.core.packages.dependency import Dependency
     from poetry.core.packages.package import Package
     from poetry.core.semver.version import Version
 
+    from poetry.repositories.repository import Repository
 
-class Pool(Repository):
+
+class RepositoryPriority(Enum):
+    # The order of the members below dictates the actual priority. The first member has
+    # top priority.
+    DEFAULT = auto()
+    PRIMARY = auto()
+    SECONDARY = auto()
+
+
+class Pool:
```

Comment:
```suggestion
class RepositoryPool:
```

Let's go all the way with breaking the false inheritance and name this clearly.

Reply:
Looking at this closer, we do have a lot of shared methods. Let's make a common `AbstractRepository` ABC that both `RepositoryPool` and `Repository` can implement.

Reply:
It turns out the number of shared methods is quite small: only `find_packages` and `search` overlap (assuming we don't put private methods in the ABC). The method `package` also exists in both, but the signature is different.

Reply:
Hmm, we could make `package` common with the use of kwargs. I'm rather inclined to do that because the idea of reasoning about these repos based on ABC instead of concrete type (when possible) is quite attractive. 

Also, I noticed for some reason `CachedRepository` does inherit from ABC without having any `@abstractmethod`s -- we should probably clean that up as well (e.g. add abstract declarations common between implementations).

Reply:
I didn't find a way to do this without `# type: ignore[override]`, but perhaps you know a better way?

`CachedRepository` has a single abstract method named `_get_release_info`.

Reply:
in your current draft, only the `AbstractRepository` has kwargs, none of the concrete classes do.  So just remove the kwargs from the abstract class, and also remove all the `# type: ignore[override]` comments?

Reply:
The `RepositoryPool` does, that's the issue..

Reply:
then add kwargs to the `RepositoryPool` only, and not to the abstract base class

Reply:
I re-reviewed the code and stand by plan A: remove the kwargs from the abstract class and remove all of the `# type: ignore[override]` comments.

Reply:
I've processed the change. I guess it is a matter of preference -- as I user I like to be informed of potential deviations from the ABC (API to me), but it did get quite ugly for little benefit.

Reply:
Alternatively, we could just ignore `kwargs` on every implementation except RepositoryPool, if that makes you feel better :laughing: